### PR TITLE
Send APNs messages on nudge PEEN server

### PR DESCRIPTION
**Pull request by bromeostasis**
Closes getlarky/nudge#113

Testing [here](https://github.com/getlarky/nudge/issues/113#issuecomment-390312215)

### CODE REVIEW

**Comment by bromeostasis** [comment]
*I want to change the tests here so they're runnable by everyone*

> Copy the NODE_ENV=test nyc --reporter=text mocha --exit $(find test -name '*.js') into a file at ./bin/test.sh

File: push/test/models/test_authentication_key.js [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        it('it should throw a validation error for any extraneous fields', (done) => {
+            let newKeyCopy = Object.assign({}, authKeyToCreate);
+            newKeyCopy.new_field = [23, 23];
+            AuthenticationKey.create(newKeyCopy, (err, insertId) => {
+                err.should.be.instanceOf(ValidationError);
+                should.not.exist(insertId);
+                done();
+            });
+        });
+    });
```
Comment (McGarveyMobile) [inline push/test/models/test_authentication_key.js]: Do you think we should include a test/validation to see if required fields are empty strings or nah?

Comment (McGarveyMobile) [reply push/test/models/test_authentication_key.js]: Also, I don't really know - are duplicates something we need to worry about/test for here?

File: push/test/models/test_authentication_key.js [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+            AuthenticationKey.create(authKeyToCreate, (err, insertId) => {
+                if (err) {return done(err);}
+                insertId.should.be.an('number');
+                apnProviders = AuthenticationKey.getApnProviders();
+                apnProviders.should.have.property(newOrganizationId);
+                done();
+            })
+        });
+
+        it('it should throw a validation error if organization id already exists in the database', (done) => {
```
Comment (McGarveyMobile) [inline push/test/models/test_authentication_key.js]: Ah, NVM RE: duplicates - I see you test for that here!

File: push/models/authenticationKey.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    key_name: {type: Sequelize.STRING, allowNull: false},
+    active: Sequelize.BOOLEAN,
+    created: Sequelize.DATE,
+    updated: Sequelize.DATE
+},
+{
+    timestamps: false
+})
+const isActive = {active: 1};
+let apnProviders = {};
```
Comment (McGarveyMobile) [inline push/models/authenticationKey.js]: I think this is just something I don't understand about apn - why store these `apnProviders` locally, instead of in a database? Won't restarting the server clear these and therefore break everything?

File: push/models/authenticationKey.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+const AuthenticationKey = sequelize.define('authentication_key', {
+    id: {type:Sequelize.INTEGER, primaryKey: true, autoIncrement: true},
+    organization_id: {type: Sequelize.INTEGER, validate: {isInt: true}, allowNull: false, unique: true},
+    key_id: {type: Sequelize.STRING, allowNull: false},
+    team_id: {type: Sequelize.STRING, allowNull: false},
+    bundle_id: {type: Sequelize.STRING, allowNull: false},
+    key_name: {type: Sequelize.STRING, allowNull: false},
+    active: Sequelize.BOOLEAN,
+    created: Sequelize.DATE,
```
[the remark this answers is not in the document]
Comment (bromeostasis) [reply push/models/authenticationKey.js]: @McGarveyMobile yep, see [here](https://github.com/getlarky/nudge-push/pull/21/files/2d032518cb590e1744d995b673b9802c284b7e38#diff-350584829d335116fb9bb42c5f422b8eR15)

File: push/models/authenticationKey.js (hunk 3 above)
Comment (bromeostasis) [reply push/models/authenticationKey.js]: > why store these apnProviders locally, instead of in a database?

Hrmm, I'm not really sure how storing/serializing entire js objects in the database would work? Just to clarify, this isn't simply key/value pairs, it's an `apn.Provider` object with a ton of configurations and methods associated. Take a look where they get created [here](https://github.com/getlarky/nudge-push/pull/21/files/2d032518cb590e1744d995b673b9802c284b7e38#diff-cd25a1d2f85d68a832deef6d64397a3cR151)

> Won't restarting the server clear these and therefore break everything?

Yep, that's why we create them all [when the server starts](https://github.com/getlarky/nudge-push/pull/21/files/2d032518cb590e1744d995b673b9802c284b7e38#diff-fcfb9379fd1785e6385a6fc57073acbdR41).

File: push/models/authenticationKey.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            return done(new ValidationError("Could not find authentication key with id " + id));
+        }
+        return done(null, authKey);
+    }).catch(err => {
+        return done(err);
+    });
+}
+
+exports.create = function(authKey, done) {
+    let extraFields = utils.getExtraneousFieldsForInput(authKey, AuthenticationKey.rawAttributes);
```
[the remark this answers is not in the document]
Comment (bromeostasis) [reply push/models/authenticationKey.js]: @McGarveyMobile nope, looks like this needs to be fixed (same for token).. @ishotjr since this is already merged, do you want me to create a new PR for this fix or a new issue to not be able to create/update these fields?

File: push/models/authenticationKey.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            } catch(err){
+                return done(new ValidationError("Key with id " + authKeys[i].id + " and name " + authKeys[i].key_name + " does not exist. Did it get created?"));
+            }
+        }
+        return done(null, apnProviders);
+    })
+
+}
+
+function deleteApnProviderForAuthKey(organization_id) {
```
[the remark this answers is not in the document]
Comment (bromeostasis) [reply push/models/authenticationKey.js]: @McGarveyMobile I see it similarly to our normal POST vs. DELETE. We pass an entire object to one vs. an identifier to the other.

I **do** see the validity of changing the confusing method name. What do you think of `deleteApnProviderByOrganizationId`?

File: push/routes/messages.js [hunk 7]
```diff
@@ -1,17 +1,45 @@
...
+    let messageId = postData.message_id
+    let env = postData.environment
+    let token = postData.token;
+
+    if (!token) {
+        token = "73ae756c3fd9a5c240c2c9e6949913aa0af3ea4e01b0fbac456c5f6bc135447e";
+    }
+    // use joi or other validation in final endpoint
+    if (env) {
+        if (env != "development" && env != "production") {
```
[the remark this answers is not in the document]
Comment (bromeostasis) [reply push/routes/messages.js]: @McGarveyMobile probably, but this code is gonna need to get re-written in getlarky/nudge#110 anyways, so I kinda figured 🤷‍♂️

File: push/test/models/test_authentication_key.js (hunk 1 above)
Comment (bromeostasis) [reply push/test/models/test_authentication_key.js]: > Do you think we should include a test/validation to see if required fields are empty strings or nah?

@McGarveyMobile yeah I think this makes sense. I'll add a few!

**Comment by bromeostasis** [comment]
@McGarveyMobile think I got everything; none of them really need your feedback, so no rush on the responses. 

@ishotjr I think I'll definitely need to make a few changes, so I'll probably just re-open the branch and do another PR once I'm done.

File: push/models/authenticationKey.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (McGarveyMobile) [reply push/models/authenticationKey.js]: @bromeostasis Ahh, yeah much better! Makes sense to me.

File: push/routes/messages.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (McGarveyMobile) [reply push/routes/messages.js]: 🤷‍♀️

File: push/models/authenticationKey.js (hunk 3 above)
Comment (McGarveyMobile) [reply push/models/authenticationKey.js]: @bromeostasis ohhhhhhhhhhh gotcha gotcha, That's what that's for! 😅

**Comment by ishotjr** [comment]
@bromeostasis oh. I was on the lookout for a PR per SU so when I saw "approved" I read it as "approved" on the assumption that comments were just to enable @McGarveyMobile's edification on all things APNs - sounds good! 🙆‍♂️

**Comment by bromeostasis** [comment]
> @bromeostasis oh. I was on the lookout for a PR per SU so when I saw "approved" I read it as "approved" on the assumption that comments were just to enable @McGarveyMobile's edification on all things APNs - sounds good! 🙆‍♂️

@ishotjr yeah that's fair; @McGarveyMobile probably should've just requested changes on this one, no need to spare my feelings 😆 👦 🥉 

> @ishotjr since this is already merged, do you want me to create a new PR for this fix on authKeys or a new issue to not be able to create/update these fields for both objects?

@ishotjr any thoughts on [this](https://github.com/getlarky/nudge-push/pull/21#discussion_r189666554)?

**Comment by ishotjr** [comment]
@bromeostasis first of all, I love `getExtraneousFieldsForInput()`!!  I vote for one new issue to fix both! :bowtie: 👍

**Comment by bromeostasis** [comment]
> @bromeostasis first of all, I love getExtraneousFieldsForInput()!! I vote for one new issue to fix both!  👍

@ishotjr getlarky/nudge#153
